Volume dumping restriction manager

samparsky opened the pull request:
### Please check if the PR fulfills these requirements
- [x] The commit message follows our Submission guidelines
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)


### What kind of change does this PR introduce? 
#268 

### What is the current behavior? 
(You can also link to an open issue here)


### What is the new behavior?
(Define and describe any new functionality. Clarify if this is a feature change)


### Does this PR introduce a breaking change?
No


### Any Other information:

Commit:
added contracts & test file

satyamakgec commented:
Hey, @samparsky Should I start reviewing this PR or still you are adding some code in it?

samparsky commented:
@satyamakgec Currently adding more tests. Will let you know when I am done

samparsky commented:
@satyamakgec Please kindly review this PR

satyamakgec commented on contracts/modules/TransferManager/VolumeDumpingRestrictionManager.sol:
```diff
@@ -0,0 +1,321 @@
+pragma solidity ^0.4.24;
+
+import "./ITransferManager.sol";
+import "../../tokens/SecurityToken.sol";
```
No requirement of importing the SecurityToken.sol here

satyamakgec commented on contracts/modules/TransferManager/VolumeDumpingRestrictionManager.sol:
```diff
@@ -0,0 +1,321 @@
+    event Testing (
+        address indexed from, 
+        uint256 text
+    );
```
Unnecessary code. - remove it

satyamakgec commented on contracts/modules/TransferManager/VolumeDumpingRestrictionManager.sol:
```diff
@@ -0,0 +1,321 @@
+     * @param _amount The amount of tokens to transfer
+     */
+    function verifyTransfer(address _from, address /*_to */, uint256 _amount, bool /* _isTransfer */) public returns(Result) {        
+        if (!paused && _from != address(0)) {                
```
Below code, you are changing the state so it means anybody can change it because `verifyTransfer()` is public function. To solve this you should apply a restriction as `require(_isTransfer == false || msg.sender == securityToken, "Sender is not owner");`. for concrete example you can take a look into the `ManualApproavalTransferManager.sol`

satyamakgec commented:
@samparsky left some comments on the PR. and you can also improve the test coverage of your contracts although it is good please try to make it `95% - 98%`

Commit:
test changes, added multi methods, tests and changed percentage calculation

samparsky commented:
@satyamakgec I have made the changes suggested above.
The test seems to be failing due to some issue I can't figure out

This returns 
```sh
$ node_modules/.bin/solidity-coverage
```
the following error which seems odd, given that `truffle compile` works perfectly
```
* Line 1, Column 1
  Syntax error: value, object or array expected.
* Line 1, Column 1
  A valid JSON document must be either an array or an object value.
```

satyamakgec commented:
Yes @samparsky we are facing this issue, Once we resolve that then your PR also start building in Travis.

satyamakgec commented on contracts/modules/TransferManager/VolumeDumpingRestrictionManager.sol:
```diff
@@ -0,0 +1,322 @@
+            uint256 allowedRemainingAmount = _currentAllowedAmount(_from, periodId);
+
+            if (_amount <= allowedRemainingAmount){
+                volumeTally[_from][periodId] = volumeTally[_from][periodId].add(_amount);
```
Here you are changing the state variable. It should be restricted and only be performed when `(_isTransfer == true or securityToken == msg.sender)`. You are allowing the change when anyone going to query the public `verifyTransfer()` function

samparsky replied:
Are there scenarios when `isTransfer` would be `false`? like a readonly txn?

satyamakgec replied:
Yes for readonly `isTransfer` should be false.

satyamakgec commented on contracts/modules/TransferManager/VolumeDumpingRestrictionManager.sol:
```diff
@@ -0,0 +1,322 @@
+                return Result.VALID;
+            }
+            
+            uint256 periodId = now.div(userDumpingRestriction.rollingPeriod);
```
I don't think it is a right approach. For correct periodID, it should be 
`uint256 periodId = (now - startTime).div(userDumpingRestriction.rollingPeriod)`

satyamakgec commented on contracts/modules/TransferManager/VolumeDumpingRestrictionManager.sol:
```diff
@@ -0,0 +1,322 @@
+
+        // check if entry is in the past
+        // not allowed to delete past entries
+        require(toRemove.endTime > now, "Cannot remove past restrictions");
```
wrong reason string. It should be `Only past restrictions are allowed to delete` more preferable is `Only passed restrictions are allowed to delete`

samparsky replied:
According to the flow chart design in the [Lucid Chart](https://www.lucidchart.com/documents/edit/c96ff64c-5b89-4b9c-985a-01637266bf88/0?shared=true&) provided It says below
![image](https://user-images.githubusercontent.com/8148384/46685907-893e0c00-cbee-11e8-9b4a-7ebd0905da1c.png)

It would contradict it its essentially checking if the restriction being removed endtime is still current. Using `Only past restrictions are allowed to delete` would require changing it to 

```javascript   
require(toRemove.endTime < now, "Only past restrictions are allowed to delete");
```

satyamakgec replied:
Hmm agreed keep as it is

samparsky commented:
@satyamakgec Please review, the changes suggested have been implemented

samparsky commented:
@satyamakgec Please review, the changes suggested have been implemented

adamdossa commented on contracts/modules/TransferManager/VolumeDumpingRestrictionManager.sol:
```diff
@@ -0,0 +1,326 @@
+            VolumeRestriction memory userDumpingRestriction = volumeRestriction[_from];
+
+            if (now < userDumpingRestriction.startTime || now > userDumpingRestriction.endTime){
+                return Result.VALID;
```
This should return Result.NA not Result.VALID (if the `_from` address doesn't have any dumping restrictions it doesn't mean the transfer is necessarily valid, just that it is not invalid from the perspective of this transfer manager)

adamdossa commented on contracts/modules/TransferManager/VolumeDumpingRestrictionManager.sol:
```diff
@@ -0,0 +1,326 @@
+
+            if (_amount <= allowedRemainingAmount){
+                volumeTally[_from][periodId] = volumeTally[_from][periodId].add(_amount);
+                return Result.VALID;
```
As above should return Result.NA here not Result.VALID

samparsky replied:
When should `Result.VALID` be used?

adamdossa commented on contracts/modules/TransferManager/VolumeDumpingRestrictionManager.sol:
```diff
@@ -0,0 +1,326 @@
+    function _currentAllowedAmount(address _userAddress, uint periodId) internal returns (uint) {
+        VolumeRestriction storage userDumpingRestriction = volumeRestriction[_userAddress];
+        
+        uint alreadyWithdrawn = volumeTally[_userAddress][periodId];
```
Can we consistently use `uint256`

adamdossa commented on contracts/modules/TransferManager/VolumeDumpingRestrictionManager.sol:
```diff
@@ -0,0 +1,326 @@
+    * @param _userAddress User Address to get the rule
+    * @param periodId Time period identifier
+    */
+    function _currentAllowedAmount(address _userAddress, uint periodId) internal returns (uint) {
```
Should be `_periodId`

adamdossa commented on contracts/modules/TransferManager/VolumeDumpingRestrictionManager.sol:
```diff
@@ -0,0 +1,326 @@
+            return Result.VALID;
+        }
+
+        if (!paused && _from != address(0)) {                
```
Probably should also check that `_amount != 0` otherwise you'll add duplicate periods in `_currentAllowedAmount` I think.

samparsky replied:
It doesn't add duplicate periodId, since I am getting the `remaining amount` allowed to be withdrawn. But I have added the check though.

adamdossa commented on contracts/modules/TransferManager/VolumeDumpingRestrictionManager.sol:
```diff
@@ -0,0 +1,326 @@
+        // add already withdrawn amount for this period 
+        // to be able to get the correct value
+
+        uint256 percentAllowed = userDumpingRestriction.percentAllowed;
```
For better accuracy percentages should be represented as [percentage] * 10**16. So 100% == 10**18, 50% = 0.5 * 10**18

samparsky commented:
@adamdossa  @satyamakgec Please review, the changes suggested have been implemented

satyamakgec commented:
@samparsky Thanks for all your hard work and patience. we are in the process of reviewing the code. it may take a couple of days to finalize the module. So please bear me. I will try to send you an update as soon as possible.

samparsky commented:
@satyamakgec @adamdossa  It would be helpful if this PR can be reviewed & merged.
Merge conflicts occurs when other PRs get merged in. I have had to rebase & fix several times cos of this, it gets frustrating after some time.

satyamakgec commented on contracts/modules/TransferManager/VolumeDumpingRestrictionManager.sol:
```diff
@@ -0,0 +1,325 @@
+        require(_isTransfer == false || msg.sender == securityToken, "Sender is not owner");
+
+        if(!_isTransfer) {
+            return Result.VALID;
```
This means it always returns valid when the `_isTrasfer` is false.  It doesn't sound well because there are higher chances to fail when the user does the actual transaction.

satyamakgec commented on contracts/modules/TransferManager/VolumeDumpingRestrictionManager.sol:
```diff
@@ -0,0 +1,325 @@
+    mapping(address => uint256[]) internal volumePeriodIds;
+    
+    event ModifyVolumeDumping (
+        address indexed from, 
```
rename the `from` variable because it sounds confusing as it looks the address which modifies the dumping restriction. So make it `restrictedAddress` or `tokenHolder`. Do the same in all other events.

satyamakgec commented on contracts/modules/TransferManager/VolumeDumpingRestrictionManager.sol:
```diff
@@ -0,0 +1,325 @@
+    * @notice lets the admin remove a volume dumping restriction.
+    * @param _userAddress User address to remove the volume dumping restriction
+    */
+    function removeRestriction(address _userAddress) public withPerm(ADMIN) {
```
Add batch version of it

satyamakgec commented on contracts/modules/TransferManager/VolumeDumpingRestrictionManager.sol:
```diff
@@ -0,0 +1,325 @@
+        require(address(0) != _userAddress, "Invalid user address");
+        require(_percentAllowed > 0 && _percentAllowed <= 100, "Invalid percent input");
+        require(_startTime == 0 || _startTime >= now, "Invalid start time");
+        require(_endTime > 0 && _endTime > _startTime, "Invalid endtime");
```
no need of `_endTime > 0` when you are checking `_endTime> _startTime` and `require(_startTime == 0 || _startTime >= now, "Invalid start time")`

satyamakgec commented on contracts/modules/TransferManager/VolumeDumpingRestrictionManager.sol:
```diff
@@ -0,0 +1,325 @@
+            _startTime = now;
+        }
+
+        require(_endTime > _startTime, "End time must be greater than start time");
```
no need of this when you already checking all values in `_checkVolumeDumpingParams()`

satyamakgec commented on contracts/modules/TransferManager/VolumeDumpingRestrictionManager.sol:
```diff
@@ -0,0 +1,325 @@
+
+        if (_startTime == 0) {
+            _startTime = now;
+        }
```
move this if statement at the top of the function before the `_checkVolumeDumpingParams()`

satyamakgec commented on contracts/modules/TransferManager/VolumeDumpingRestrictionManager.sol:
```diff
@@ -0,0 +1,325 @@
+    * @param _endTime When the dumping restriction ends in seconds
+    * @param _rollingPeriod is the time period in seconds
+    */
+    function modifyVolumeDumpingRestriction( 
```
Add a batch version of this function i.e - `modifyVolDumpingRestricionMulti()`

satyamakgec commented on contracts/modules/TransferManager/VolumeDumpingRestrictionManager.sol:
```diff
@@ -0,0 +1,325 @@
+        // add already withdrawn amount for this period 
+        // to be able to get the correct value
+
+        uint256 percentAllowed = userDumpingRestriction.percentAllowed.mul(uint256(10)**16);
```
This multiplication should not be here. it should check at the addition time otherwise issuer not be able to put the percentage in decimals ex- 12.5 %, 35.25% etc.

samparsky replied:
Does this mean the percentage values passed in `addRestriction`would be in the format e.g.
12 % = 12 * 10 ** 16
12.5 = 125 * 10 ** 15

satyamakgec replied:
Yes, the value will always between 1 * 10^16 - 100 * 10^16.
12.5 = 125 * 10^15

satyamakgec commented on contracts/modules/TransferManager/VolumeDumpingRestrictionManager.sol:
```diff
@@ -0,0 +1,325 @@
+        uint256 percentAllowed = userDumpingRestriction.percentAllowed.mul(uint256(10)**16);
+
+        // totalAmountAllowedForPeriod = currentbalance * 20/100
+        uint256 totalAmountAllowedForPeriod = currentUserBalance.add(alreadyWithdrawn).mul(percentAllowed).div(100 * uint256(10)**16);
```
This will give you a wrong calculation 
ex- A will have the balance of 1000 tokens 
      percentage allowed - 10%  = 100 tokens 
      First txn deduct the 50 tokens and after first txn A will get the 1000 tokens from B.
 for the second transaction(same periodId) calculation will look like this 
    (1950 + 50 ) * (10)/100 -- it is totally wrong because now it will allow him to deduct 200 tokens instead of more 50 tokens.

samparsky replied:
This would be violating the design document shared with me. According to https://www.lucidchart.com/documents/edit/c96ff64c-5b89-4b9c-985a-01637266bf88/0?shared=true&

![image](https://user-images.githubusercontent.com/8148384/47265000-151e3500-d519-11e8-9a05-fdbd6951e592.png)

samparsky replied:
@satyamakgec Can you help confirm this?

satyamakgec commented on contracts/modules/TransferManager/VolumeDumpingRestrictionManagerFactory.sol:
```diff
@@ -0,0 +1,102 @@
+        version = "1.0.0";
+        name = "VolumeDumpingTransferManager";
+        title = "Volume Dumping Transfer Manager";
+        description = "Manage the volume of tokens dumpable by an owner";
```
add the compatible protocol version range. Please check the other module factories you can easily understand what I am trying to say

satyamakgec commented on contracts/modules/TransferManager/VolumeDumpingRestrictionManager.sol:
```diff
@@ -0,0 +1,325 @@
+ * @title  Transfer Manager module for restricting volume dumping within a time period for an account
+ */
+
+contract VolumeDumpingRestrictionManager is ITransferManager {
```
change the contract name to `VolumeDumpingRstrictionTM`

satyamakgec commented on migrations/2_deploy_contracts.js:
```diff
@@ -9,6 +9,7 @@ const ERC20DividendCheckpointFactory = artifacts.require('./ERC20DividendCheckpo
 const ModuleRegistry = artifacts.require('./ModuleRegistry.sol');
 const ModuleRegistryProxy = artifacts.require('./ModuleRegistryProxy.sol');
 const ManualApprovalTransferManagerFactory = artifacts.require('./ManualApprovalTransferManagerFactory.sol')
+const VolumeDumpingRestrictionTransferManagerFactory = artifacts.require('./VolumeDumpingRestrictionManagerFactory.sol')
```
Please remove from the migrations

satyamakgec commented on migrations/2_deploy_contracts.js:
```diff
@@ -161,6 +162,10 @@ module.exports = function (deployer, network, accounts) {
+  }).then(() => {
+    // D) Deploy the VolumeDumpingRestrictionTransferManagerFactory Contract (Factory used to generate the VolumeDumpingRestrictionTransferManager contract use
+    // to track the dumping crietria an investor could do for a particular security token)
+    return deployer.deploy(VolumeDumpingRestrictionTransferManagerFactory, PolyToken, 0, 0, 0, {from: PolymathAccount});
```
Please remove from the migrations

satyamakgec commented on scripts/test.sh:
```diff
@@ -63,7 +63,7 @@ start_testrpc() {
     node_modules/.bin/testrpc-sc --gasLimit 0xfffffffffff --port "$testrpc_port" "${accounts[@]}" > /dev/null &
   else
-    node_modules/.bin/ganache-cli --gasLimit 8000000 "${accounts[@]}" > /dev/null &
+    node_modules/.bin/ganache-cli --gasLimit 0xfffffffffff "${accounts[@]}" > /dev/null &
```
We want to run all the txn within the Mainnet gas range i.e 8 MN

satyamakgec reviewed (changes requested):
Also, try to increase the branch coverage it is 80 % but we are expecting b/w 95% - 100%

Commit:
remove uneeded changes

Commit:
remove uneeded changes

Commit:
added more tests

Commit:
fixed a test

Commit:
added more tests

added more tests

added movePeriod amounts

added changes

samparsky commented:
@satyamakgec I have implemented the changes you have suggested. 
The code coverage for the files added is at 96% & 98%.

VolumeDumpingRestrictionTMFactory.sol - 96%
VolumeDumpingRestrictionTM.sol - 98%


I still have some questions,
Regarding this below, I haven't gotten any clear response from you yet as it doesn't quite agree with the design I was given https://www.lucidchart.com/documents/edit/c96ff64c-5b89-4b9c-985a-01637266bf88/0?shared=true&

```
This will give you a wrong calculation
ex- A will have the balance of 1000 tokens
percentage allowed - 10% = 100 tokens
First txn deduct the 50 tokens and after first txn A will get the 1000 tokens from B.
for the second transaction(same periodId) calculation will look like this
(1950 + 50 ) * (10)/100 -- it is totally wrong because now it will allow him to deduct 200 tokens instead of more 50 tokens.
```
It would be awesome to get a response

satyamakgec commented:
@samparsky Thanks for the changes Yes you are right it is not matching with the design. I will give you more information after the internal discussion on design.

samparsky commented:
@satyamakgec @adamdossa  I am waiting on your decision

satyamakgec commented:
Hey @samparsky sorry for the delay. The logic looks as - 
let’s assume the limit is 10,000 tokens in a rolling 30 days. If SUM (all sales in the last 29 days) + sale requested now < 10,000 then the transaction can proceed, else the transaction reverts. and yes you are doing right as per the lucidchart. Always takes the user current balance in the calculation. But please keep in mind the logic of rolling period which seems off to me in your current code.

samparsky commented:
@satyamakgec That means the calculation of the limit  is correct,

As for the logic of the rolling period is calculated thus

The assumption is there is no rollover of tokens transfer i.e. if you have 10k limit in a period & you only transfer 5k you don't get the remaining 5k rolled over to the next period

endtime - 200
starttime - 100
rollingperiod - 10

currentTime = 101
unit256 periodId = (101-100) / 10 = 0

currentTime = 102
uint256 periodId = (102 -100) / 10 = 0

currentTime = 120
uint256 periodId = (120-100) / 10 = 2

currentTime = 150
uint256 periodId = (150-100) / 10 = 5

It would be great if we can work together to close the PR asap, its taken quite a very long time.

samparsky commented:
@satyamakgec @adamdossa Still waiting on you

satyamakgec commented:
Hey we are at Devcon so the coming week will be pretty much busy. I will
catch you up after the devcon.
Cheers

On Mon, Oct 29, 2018, 12:25 Omidiora Samuel <notifications@github.com>
wrote:

> @satyamakgec <https://github.com/satyamakgec> @adamdossa
> <https://github.com/adamdossa> Still waiting on you
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/PolymathNetwork/polymath-core/pull/292#issuecomment-433876046>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AN5_BVgFNplBi81eBBsugxyI20QexsSaks5upuW4gaJpZM4W_Q-U>
> .
>

samparsky commented:
@satyamakgec Hello, how was Devcon? Hoping we can conclude on this, this week

satyamakgec commented:
@samparsky Devcon was good and yes we can conclude this week.

satyamakgec commented on contracts/modules/TransferManager/VolumeDumpingRestrictionTM.sol:
```diff
@@ -0,0 +1,366 @@
+
+        // check if entry is in the past
+        // not allowed to delete past entries
+        require(toRemove.endTime > now, "Cannot remove past restrictions");
```
It doesn't make sense to delete the past entries. If issuer accidentally add someone then no way to remove the restriction.

samparsky replied:
Yeah, that's what is implemented.

If the `endTime > now`, this means 
i.e. endTime = 150 & currentTime = 100.
This means that the restriction is still current so you can remove it but once it's not
i.e. endTime = 150 & currentTime = 200
Then the restriction is in the past and you can't modify it

samparsky replied:
`If issuer accidentally add someone then no way to remove the restriction.` As far has the endTime hasn't been exceeded the admin can remove the restriction.

satyamakgec replied:
Sorry My bad, I misread

satyamakgec commented on contracts/modules/TransferManager/VolumeDumpingRestrictionTM.sol:
```diff
@@ -0,0 +1,366 @@
+
+        _checkVolumeDumpingParams(_userAddress, _percentAllowed, _startTime, _endTime, _rollingPeriod);
+ 
+        require(volumeRestriction[_userAddress].endTime > now, "Cannot modify past restrictions");
```
same if the issuer wants to modify the details for a particular investor. He can't do it until the dumping period gets completed. But please do care with your logic because if Issuer changes details during the dumping period then it will affect the calculation of rolling period. 
Two ways to mitigate - 
1. Allow the ADMIN to modify only when startTime > now.
2. Change the logic of periodID to make it more dynamic.

samparsky replied:
I included that if the startTime hasn't been reached then you can modify the start time else you can't.

Can we clarify using the below example

For example if a user has the following config 
percentAllowed = 10
rollingPeriod = 30

And then intends to change it to
percentAllowed = 20
rollingPeriod = 30

It doesn't really affect anything

But if he changes it to
percentAllowed = 10
rollingPeriod = 40

If the rolling period has started then definitely it would affect the calculation, since the rolling period & start time is used in calculating the period id, 

Hence an approach would be to check the already spent amount in the current active auction then roll over the spent amount to the new periodId calculated.

I also would have to delete all previously existing periodIds to prevent a case of periodId collision

That would seem to take care of the edge case

Or is there anything else I am missing?

satyamakgec replied:
Yes, you are covering all the base but I am not a fan of deleting the previous data because if any reason Issuer wants to get the details from the contract then Issuer receives fewer data.

samparsky replied:
Yeah another approach would be to delete the only the periodIds greater than the newPeriodID, to prevent the id collision

samparsky replied:
Or we can prevent changing the rollingPeriod & startTime since that's what is used in calculating the periodId once the restriction period has started
So if any need arises to modify the rolling period the issuer & the user would have to use a new address.

satyamakgec commented:
@samparsky please do resolve the above issues and update your branch so I can give this module a green light

samparsky committed:
modified volume dumping

samparsky committed:
test changes

samparsky committed:
test changes

samparsky committed:
test changes

samparsky committed:
fix case

samparsky committed:
added more test cases

samparsky commented:
@satyamakgec I have implemented the suggested changes.  I changed the implementation of the `periodId` to be more dynamic to allow changing the rollingPeriod without affecting the periodId calculation.

maxsam4 commented:
Bounty paid. Thanks a lot for your contribution @samparsky  .

samparsky commented:
@maxsam4 Any reason why the PR wasn't merged?

maxsam4 commented:
@samparsky We have decided to merge this module and https://github.com/PolymathNetwork/polymath-core/pull/320

It would not have been fair of us to ask you guys to do the required change so we'll be doing them ourselves. 
If you want to work on merging them in your free time, feel free and I'll reopen this PR. Thank you for all your effort, we really appreciate your work.
